README.md

# Crumbs

> Turn asterisk-indented text lines into mind maps.

Organize your notes in a hierarchical tree structure, using a simple text editor.

- an asterisk at the beginning of the line means _level 1_
- two asterisk at the beginning of the line means _level 2_
- and so on...

# How [crumbs](https://github.com/lucasepe/crumbs/releases/latest) works?

- takes in input a text file and generates a [dot script](https://en.wikipedia.org/wiki/DOT_(graph_description_language)) for [Graphviz](https://graphviz.gitlab.io/download/).

- depends on [GraphViz](https://graphviz.gitlab.io/download/)
  - look at the bottom for info about [how to install graphviz](#how-to-install-graphViz).


## Example (without icons)

Create a simple text file - let's say `meeting-ideas.txt`:

```text
* main idea
** topic 1
*** sub topic
*** sub topic
**** sub topic
**** sub topic
** topic 2
*** sub topic
```

To create the PNG image you can 

- _"inject"_ the text file to [crumbs](https://github.com/lucasepe/crumbs/releases/latest) and then the result to [dot](https://graphviz.org/doc/info/command.html) redirecting the output to the file `meeting-ideas.png` - (I love the Linux command pipelines! 😍)

```bash
cat meeting-ideas.txt | crumbs | dot -Tpng > meeting-ideas.png
```

- or as alternative you can specify your text file to [crumbs](https://github.com/lucasepe/crumbs/releases/latest) directly:

```bash
crumbs meeting-ideas.txt | dot -Tpng > meeting-ideas.png
```

Here the output:

![](./testdata/sample4.png)

---

## Example (with icons)

You can, eventually, add images too (one for text line) using a special syntax: `[[path/to/image.png]]`

- if you specify the flag `-image-path` you can write `[[image.png]]` instead of `[[path/to/image.png]]`
- if you specify the flag `-image-type` you can write `[[path/to/image]]` instead of `[[path/to/image.png]]`
- therefore if you specify both you can write `[[image]]` instead of `[[path/to/image.png]]`

```text
* [[./png/bulb.png]] main idea
** topic 1
*** sub topic
*** sub topic
**** [[./png/comments-alt.png]] sub topic
**** sub topic
** [[./png/map-signs.png]] topic 2
*** sub topic
```

then as usual, let's feed graphviz with [crumbs](https://github.com/lucasepe/crumbs/releases/latest):

```bash
crumbs meeting-ideas-with-icons.txt | dot -Tpng > meeting-ideas-with-icons.png
```

and the output is...

![](./testdata/sample5.png)

## Example (with HTML)

You can enrich the output with a little bit of style, adding some HTML tag.

The following tags are understood:

```html
<b>, <br/>, <i>, <o>, <s>, <sub>, <sup>, <u>
```

```text
* main idea
** <u>topic 1</u>
*** sub <sub>topic</sub>
*** sub <i>topic</i>
**** <s>sub topic</s>
**** sub <o>topic</o>
** topic <b>2</b>
*** sub <sup>topic</sup>
```

then as usual, let's feed graphviz with [crumbs](https://github.com/lucasepe/crumbs/releases/latest):

```bash
crumbs meeting-ideas-with-html.txt | dot -Tpng > meeting-ideas-with-html.png
```

and the output is...

![](./testdata/sample6.png)

---

# Installation Steps

In order to use the crumbs command, compile it using the following command:

```bash
go get -u github.com/lucasepe/crumbs/crumbs
```

This will create the crumbs executable under your $GOPATH/bin directory.


## Ready-To-Use Releases 

If you don't want to compile the sourcecode yourself, [Here you can find the tool already compiled](https://github.com/lucasepe/crumbs/releases/latest) for:

- MacOS
- Linux
- Windows

---

# CHANGE LOG

👉 [Record of all notable changes made to a project](./CHANGELOG.md)

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.3.0] - 2020-11-09
### Added
- 🎉 new flag `-images-type` to specify a default suffix for all the images
  - specifying this flag when including images, you can omit the extension 
  - example: if flag has `-images-type png`, you can write `[[bulb]]` instead of `[[build.png]]`

### Changed
- updated the README markdown file
- improve argument parsing
  - when invoked without args, an attempt is made to read from standard input

### Fixed
- 🐛 Line with no leading stars causes nil pointer dereference [#2](/issues/#2) 

## [0.2.0] - 2020-09-09
### Added
- 📝 more test cases
- 🎉 new flag `images-path` to specify the base images folder
  - now when including images, you can specify just the filename
  - example: if flag has `-images-path '/Icons/AwesomFonts'`, you can write `[[bulb.png]]` instead of `[[/Icons/AwesomFonts/build.png]]`

- 🎉 new flag `lim` to specify after how many characters to wrap the text

### Changed
- updated the README markdown file

### Fixed
- 🐛 removed `go.sum` from the .gitignore file 

## [0.1.0] - 2020-09-08
- 🎉 First release!


parser_test.go
package crumbs

import (
	"strings"
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestParseLines(t *testing.T) {
	test := `
* main idea
** topic 1
*** sub topic 1 1
*** sub topic 1 2
**** sub sub topic
** topic 2
*** sub topic 2 1
`
	got, err := ParseLines(strings.SplitAfter(test, "\n"), "")
	if err != nil {
		t.Error(err)
	}

	assert.Equal(t, 1, len(got.childrens))
	assert.Equal(t, got.childrens[0].text, "main idea")

	assert.Equal(t, 2, len(got.childrens[0].childrens))
	assert.Equal(t, got.childrens[0].childrens[0].text, "topic 1")
	assert.Equal(t, got.childrens[0].childrens[1].text, "topic 2")

	assert.Equal(t, 2, len(got.childrens[0].childrens[0].childrens))
	assert.Equal(t, got.childrens[0].childrens[0].childrens[1].text, "sub topic 1 2")
}

func TestLookForIcon(t *testing.T) {

	tests := []struct {
		imagespath string
		entry      Entry
		want       string
	}{
		{
			"./images/png",
			Entry{text: "[[blob.png]] La vispa Teresa avea tra l'erbetta, a volo sorpresa"},
			"images/png/blob.png",
		},
		{
			"/home/lus/Pictures/fontawesome/PNG",
			Entry{text: "[[blob.png]] La vispa Teresa avea tra l'erbetta, a volo sorpresa"},
			"/home/lus/Pictures/fontawesome/PNG/blob.png",
		},
	}

	for _, tt := range tests {
		fn := lookForIcon(tt.imagespath)
		fn(&tt.entry)

		t.Run(tt.imagespath, func(t *testing.T) {
			if got := tt.entry.icon; got != tt.want {
				t.Errorf("got [%v] want [%v]", got, tt.want)
			}
		})
	}

}

func TestDepth(t *testing.T) {
	tests := []struct {
		line string
		want int
	}{
		{"* main idea LV.1", 1},
		{"** topic 1 LV.2", 2},
		{"*** sub topic LV.3", 3},
		{"******* LV.7", 7},
	}

	for _, tt := range tests {

		t.Run(tt.line, func(t *testing.T) {
			assert.Equal(t, depth(tt.line), tt.want)
		})
	}
}


gv/graph_test.go
package gv

import (
	"fmt"
	"strings"
	"testing"

	"github.com/emicklei/dot"
)

func TestNewNodeOptions(t *testing.T) {
	tests := []struct {
		opts []nodeAttribute
		want string
	}{
		{
			[]nodeAttribute{},
			`digraph  {n1[fontname="Fira Code",fontsize="12",label="",margin="0.2,0.2",width="2"];}`,
		},

		{
			[]nodeAttribute{nodeFillColor("#ff0000")},
			`digraph  {n1[fillcolor="#ff0000",fontname="Fira Code",fontsize="12",label="",margin="0.2,0.2",style="filled",width="2"];}`,
		},

		{
			[]nodeAttribute{nodeShape("box")},
			`digraph  {n1[fontname="Fira Code",fontsize="12",label="",margin="0.2,0.2",shape="box",width="2"];}`,
		},

		{
			[]nodeAttribute{nodeShape("hexagon"), nodeFillColor("#00ff00")},
			`digraph  {n1[fillcolor="#00ff00",fontname="Fira Code",fontsize="12",label="",margin="0.2,0.2",shape="hexagon",style="filled",width="2"];}`,
		},
	}

	for i, tt := range tests {
		t.Run(fmt.Sprintf("test_%d", i), func(t *testing.T) {
			gv := dot.NewGraph()
			createNode(gv, "ID", tt.opts...)
			if got := flatten(gv.String()); got != tt.want {
				t.Errorf("got [%v] want [%v]", got, tt.want)
			}
		})
	}
}

func TestNodeHTMLabel(t *testing.T) {
	tests := []struct {
		html string
		want string
	}{
		{
			`<b>Bold</b>`,
			`digraph  {n1[fontname="Fira Code",fontsize="12",label=<<b>Bold</b>>,margin="0.2,0.2",width="2"];}`,
		},
		{
			`<table><tr><td>col 1</td></tr></table>`,
			`digraph  {n1[fontname="Fira Code",fontsize="12",label=<<table><tr><td>col 1</td></tr></table>>,margin="0.2,0.2",width="2"];}`,
		},
	}

	for i, tt := range tests {
		t.Run(fmt.Sprintf("test_%d", i), func(t *testing.T) {
			gv := dot.NewGraph()
			createNode(gv, "ID", nodeLabel(tt.html, true))
			if got := flatten(gv.String()); got != tt.want {
				t.Errorf("got [%v] want [%v]", got, tt.want)
			}
		})
	}
}

// remove tabs and newlines and spaces
func flatten(s string) string {
	return strings.Replace((strings.Replace(s, "\n", "", -1)), "\t", "", -1)
}


text/wrap_test.go
package text

import (
	"testing"
)

func TestWrapString(t *testing.T) {
	cases := []struct {
		Input, Output string
		Lim           uint
	}{
		// A simple word passes through.
		{
			"foo",
			"foo",
			4,
		},
		// A single word that is too long passes through.
		// We do not break words.
		{
			"foobarbaz",
			"foobarbaz",
			4,
		},
		// Lines are broken at whitespace.
		{
			"foo bar baz",
			"foo\nbar\nbaz",
			4,
		},
		// Lines are broken at whitespace, even if words
		// are too long. We do not break words.
		{
			"foo bars bazzes",
			"foo\nbars\nbazzes",
			4,
		},
		// A word that would run beyond the width is wrapped.
		{
			"fo sop",
			"fo\nsop",
			4,
		},
		// Whitespace that trails a line and fits the width
		// passes through, as does whitespace prefixing an
		// explicit line break. A tab counts as one character.
		{
			"foo\nb\t r\n baz",
			"foo\nb\t r\n baz",
			4,
		},
		// Trailing whitespace is removed if it doesn't fit the width.
		// Runs of whitespace on which a line is broken are removed.
		{
			"foo    \nb   ar   ",
			"foo\nb\nar",
			4,
		},
		// An explicit line break at the end of the input is preserved.
		{
			"foo bar baz\n",
			"foo\nbar\nbaz\n",
			4,
		},
		// Explicit break are always preserved.
		{
			"\nfoo bar\n\n\nbaz\n",
			"\nfoo\nbar\n\n\nbaz\n",
			4,
		},
		// Complete example:
		{
			" This is a list: \n\n\t* foo\n\t* bar\n\n\n\t* baz  \nBAM    ",
			" This\nis a\nlist: \n\n\t* foo\n\t* bar\n\n\n\t* baz\nBAM",
			6,
		},
	}

	for i, tc := range cases {
		actual := WrapString(tc.Input, tc.Lim)
		if actual != tc.Output {
			t.Fatalf("Case %d Input:\n\n`%s`\n\nActual Output:\n\n`%s`", i, tc.Input, actual)
		}
	}
}


note.go
package crumbs

// Entry is a thought, an idea.
type Entry struct {
	id        string
	level     int
	text      string
	icon      string
	parent    *Entry
	childrens []*Entry
}

// ID returns the node identifier.
func (ti *Entry) ID() string {
	return ti.id
}

// Childrens returns the node childs.
func (ti *Entry) Childrens() []*Entry {
	return ti.childrens
}

// Text returns the node data.
func (ti *Entry) Text() string {
	return ti.text
}

// Icon returns the icon path.
func (ti *Entry) Icon() string {
	return ti.icon
}

// Level returns the node depth.
func (ti *Entry) Level() int {
	return ti.level
}

// Parent returns the node parent
func (ti *Entry) Parent() *Entry {
	return ti.parent
}

// Root returns the root note
func (ti *Entry) Root() *Entry {
	if ti.parent == nil {
		return ti
	}
	return ti.parent.Root()
}


parser.go
package crumbs

import (
	"fmt"
	"path/filepath"
	"regexp"
	"strings"

	"github.com/teris-io/shortid"
)

// ParseLines parses a slice of text lines and builds the tree.
func ParseLines(lines []string, imagesPath, imagesSuffix string) (*Entry, error) {
	mkID := idGenerator()
	checkIcon := lookForIcon(imagesPath, imagesSuffix)

	// generate a short id for the root node
	rootID, err := mkID()
	if err != nil {
		return nil, err
	}

	// create the root node
	root := newEmptyNote(rootID)

	node := root
	nodeDepth := 0
	for _, el := range lines {
		// skip empty lines
		if strings.TrimSpace(el) == "" {
			continue
		}

		// count depth
		childDepth := depth(el)

		// case: no leading 'stars' (skip line)
		if childDepth == 0 {
			continue
		}

		// trim leading 'stars', then the spaces
		text := el[childDepth:]
		text = strings.TrimSpace(text)

		// create the child
		childID, err := mkID()
		if err != nil {
			return nil, err
		}
		child := newNote(childID, childDepth, text)
		// check if has an icon
		checkIcon(child)

		// case: the current 'node' is the parent
		if childDepth > nodeDepth {
			// update tree
			child.parent = node
			node.childrens = append(node.childrens, child)

			// update loop state
			node = child
			nodeDepth++

			// case: the current 'node' is not the parent of our child
		} else if childDepth <= nodeDepth {
			// adjust 'node' until it's correct
			for childDepth <= nodeDepth {
				node = node.parent
				nodeDepth--
			}

			// update tree
			child.parent = node
			node.childrens = append(node.childrens, child)

			// update loop state
			node = child
			nodeDepth++
		}
	}

	return root, nil
}

// depth space-counting helper (probably done in a dumb way, dunno)
func depth(line string) int {
	i := 0
	for line[i] == '*' {
		i++
	}

	return i
}

// newNote creates a new note element
func newNote(id string, lvl int, txt string) *Entry {
	f := new(Entry)
	f.id = id
	f.text = txt
	f.level = lvl
	return f
}

// newNote creates a new note element
func newEmptyNote(id string) *Entry {
	f := new(Entry)
	f.id = id
	f.level = -1
	return f
}

// idGenerator generates a new short id at each invocation.
func idGenerator() func() (string, error) {
	sid, err := shortid.New(1, shortid.DefaultABC, 2342)
	if err != nil {
		return func() (string, error) {
			return "", err
		}
	}

	return func() (string, error) {
		return sid.Generate()
	}
}

func lookForIcon(imagesPath, imagesSuffix string) func(note *Entry) {
	re := regexp.MustCompile(`^\[{2}(.*?)\]{2}`)

	return func(note *Entry) {
		str := note.text
		res := re.FindStringSubmatch(str)
		if len(res) > 0 {
			note.icon = filepath.Join(imagesPath, strings.TrimSpace(res[1]))
			if len(imagesSuffix) > 0 {
				note.icon = fmt.Sprintf("%s.%s", note.icon, imagesSuffix)
			}
			note.text = re.ReplaceAllString(str, "")
		}
	}
}


gv/render.go
package gv

import (
	"fmt"
	"io"
	"strings"

	"github.com/emicklei/dot"
	"github.com/lucasepe/crumbs"
	"github.com/lucasepe/crumbs/text"
)

// RenderConfig defines some render parameters.
type RenderConfig struct {
	VerticalLayout bool
	WrapTextLimit  uint
}

// Render translates the mind note tree to a
// graphviz dot language definition.
func Render(wr io.Writer, note *crumbs.Entry, cfg RenderConfig) error {
	htmlize := htmlLabelMaker(cfg.WrapTextLimit)

	gr := newGraph(Vertical(cfg.VerticalLayout))

	renderTree(gr, note.Root(), htmlize)

	_, err := io.WriteString(wr, gr.String())
	return err
}

// render a tree node (the node, and its children)
func renderTree(gr *dot.Graph, el *crumbs.Entry, htmlize func(*crumbs.Entry) string) {
	tintFor := colorSupplier()

	if el.Level() > 0 {
		createNode(gr, el.ID(), nodeLabel(htmlize(el), true))
	}

	if el.Parent() != nil {
		createEdge(gr, el.Parent().ID(), el.ID(), tintFor(el.Level()))
	}

	for _, child := range el.Childrens() {
		renderTree(gr, child, htmlize)
	}
}

func colorSupplier() func(lvl int) string {
	palette := map[int]string{
		0: "#264653",
		1: "#2A9D8F",
		2: "#E9C46A",
		3: "#E76F51",
		4: "#FFCDB2",
		5: "#B5838D",
		6: "#6D6875",
	}

	return func(lvl int) string {
		if val, ok := palette[lvl]; ok {
			return val
		}
		return "#000000"
	}
}

func htmlLabelMaker(lim uint) func(*crumbs.Entry) string {
	escaper := strings.NewReplacer(
		`&`, "&amp;",
		`'`, "&#39;",
		`"`, "&#34;",
	)

	return func(note *crumbs.Entry) string {
		label := strings.TrimSpace(note.Text())
		if lim > 0 {
			label = text.WrapString(label, lim)
		}
		label = escaper.Replace(label)
		label = strings.ReplaceAll(label, "\n", "<br/>")

		var sb strings.Builder
		sb.WriteString(`<table border="0" cellborder="0">`)

		if len(note.Icon()) > 0 {
			sb.WriteString("<tr>")
			fmt.Fprintf(&sb, `<td fixedsize="true" width="48" height="48"><img src="%s" /></td>`, note.Icon())
			sb.WriteString("</tr>")
		}

		switch {
		case note.Level() == 1:
			fmt.Fprintf(&sb, `<tr><td><font point-size="14"><b>%s</b></font></td></tr>`, label)
		case note.Level() > 1:
			fmt.Fprintf(&sb, `<tr><td><font point-size="12">%s</font></td></tr>`, label)
		}

		sb.WriteString("</table>")

		return sb.String()
	}
}


text/wrap.go
package text

import (
	"bytes"
	"unicode"
)

// WrapString wraps the given string within lim width in characters.
//
// Wrapping is currently naive and only happens at white-space. A future
// version of the library will implement smarter wrapping. This means that
// pathological cases can dramatically reach past the limit, such as a very
// long word.
func WrapString(s string, lim uint) string {
	// Initialize a buffer with a slightly larger size to account for breaks
	init := make([]byte, 0, len(s))
	buf := bytes.NewBuffer(init)

	var current uint
	var wordBuf, spaceBuf bytes.Buffer

	for _, char := range s {
		if char == '\n' {
			if wordBuf.Len() == 0 {
				if current+uint(spaceBuf.Len()) > lim {
					current = 0
				} else {
					current += uint(spaceBuf.Len())
					spaceBuf.WriteTo(buf)
				}
				spaceBuf.Reset()
			} else {
				current += uint(spaceBuf.Len() + wordBuf.Len())
				spaceBuf.WriteTo(buf)
				spaceBuf.Reset()
				wordBuf.WriteTo(buf)
				wordBuf.Reset()
			}
			buf.WriteRune(char)
			current = 0
		} else if unicode.IsSpace(char) {
			if spaceBuf.Len() == 0 || wordBuf.Len() > 0 {
				current += uint(spaceBuf.Len() + wordBuf.Len())
				spaceBuf.WriteTo(buf)
				spaceBuf.Reset()
				wordBuf.WriteTo(buf)
				wordBuf.Reset()
			}

			spaceBuf.WriteRune(char)
		} else {

			wordBuf.WriteRune(char)

			if current+uint(spaceBuf.Len()+wordBuf.Len()) > lim && uint(wordBuf.Len()) < lim {
				buf.WriteRune('\n')
				current = 0
				spaceBuf.Reset()
			}
		}
	}

	if wordBuf.Len() == 0 {
		if current+uint(spaceBuf.Len()) <= lim {
			spaceBuf.WriteTo(buf)
		}
	} else {
		spaceBuf.WriteTo(buf)
		wordBuf.WriteTo(buf)
	}

	return buf.String()
}


crumbs/main.go
package main

import (
	"flag"
	"fmt"
	"io"
	"io/ioutil"
	"os"
	"path/filepath"
	"strings"

	"github.com/lucasepe/crumbs"
	"github.com/lucasepe/crumbs/gv"
)

const (
	maxFileSize int64 = 512 * 1000 // 512 Kb
	banner            = `
    __  ____  __ __  ___ ___  ____    _____
   /  ]|    \|  |  ||   |   ||    \  / ___/ v{{VERSION}}
  /  / |  D  )  |  || _   _ ||  o  )(   \_ 
 /  /  |    /|  |  ||  \_/  ||     | \__  |
/   \_ |    \|  :  ||   |   ||  O  | /  \ |
\     ||  .  \     ||   |   ||     | \    |
 \____||__|\_|\__,_||___|___||_____|  \___| 
Crafted with passion by Luca Sepe - https://github.com/lucasepe/crumbs`
)

var (
	version = "0.3.0"

	flagVertical   bool
	flagWrapLim    uint
	flagImagesPath string
	flagImagesType string
)

func main() {
	configureFlags()

	entry, err := readEntry()
	if err != nil {
		fmt.Fprintf(os.Stderr, "error: %s\n", err.Error())
		os.Exit(1)
	}

	cfg := gv.RenderConfig{
		WrapTextLimit:  flagWrapLim,
		VerticalLayout: flagVertical,
	}
	if err = gv.Render(os.Stdout, entry, cfg); err != nil {
		fmt.Fprintf(os.Stderr, "error: %s\n", err.Error())
		os.Exit(1)
	}
}

func readInput() ([]byte, error) {
	limit := maxFileSize
	args := flag.Args()
	if len(args) == 0 {
		return readFileObject(os.Stdin, limit)
	}
	return readFile(args[0], limit)
}

func readEntry() (*crumbs.Entry, error) {
	src, err := readInput()
	if err != nil {
		return nil, err
	}
	text := string(src)
	lines := strings.SplitAfter(text, "\n")
	return crumbs.ParseLines(lines, flagImagesPath, flagImagesType)
}

func readFileObject(r io.Reader, limit int64) ([]byte, error) {
	lr := io.LimitReader(r, limit)
	return ioutil.ReadAll(lr)
}

func readFile(name string, limit int64) ([]byte, error) {
	r, err := os.Open(name)
	if err != nil {
		return nil, err
	}
	defer r.Close()

	return readFileObject(r, limit)
}

func configureFlags() {
	name := appName()

	flag.CommandLine.Usage = func() {
		printBanner()
		fmt.Printf("Turn asterisk-indented text lines into mind maps.\n\n")

		fmt.Print("USAGE:\n\n")
		fmt.Printf("  %s [flags] <path/to/your/file.txt>\n\n", name)

		fmt.Print("EXAMPLE(s):\n\n")
		fmt.Printf("  %s agenda.txt | dot -Tpng > output.png\n", name)
		fmt.Printf("  cat agenda.txt | %s | dot -Tpng > output.png\n\n", name)

		fmt.Print("FLAGS:\n\n")
		flag.CommandLine.SetOutput(os.Stdout)
		flag.CommandLine.PrintDefaults()
		flag.CommandLine.SetOutput(ioutil.Discard) // hide flag errors
		fmt.Print("  -help\n\tprints this message\n")
		fmt.Println()
	}

	flag.CommandLine.SetOutput(ioutil.Discard) // hide flag errors
	flag.CommandLine.Init(os.Args[0], flag.ExitOnError)

	flag.CommandLine.BoolVar(&flagVertical, "vertical", false,
		"layout entries as vertical directed graph")
	flag.CommandLine.UintVar(&flagWrapLim, "lim", 28, "wraps each line within this width in characters")

	flag.CommandLine.StringVar(&flagImagesPath, "images-path", "", "folder in which to look for image files")
	flag.CommandLine.StringVar(&flagImagesType, "images-type", "", "images file extension [png,jpg,svg]")

	flag.CommandLine.Parse(os.Args[1:])
}

func printBanner() {
	str := strings.Replace(banner, "{{VERSION}}", version, 1)
	fmt.Print(str, "\n\n")
}

func appName() string {
	return filepath.Base(os.Args[0])
}

// exitOnErr check for an error and eventually exit
func exitOnErr(err error) {
	if err != nil {
		fmt.Fprintf(os.Stderr, "error: %s\n", err.Error())
		os.Exit(1)
	}
}


gv/graph.go
package gv

import (
	"fmt"
	"strconv"
	"strings"

	"github.com/emicklei/dot"
	"github.com/lucasepe/crumbs/text"
)

// GraphOption is a graph attribute.
type GraphOption func(*dot.Graph)

// ImagesPath specifies a list of directories in which
// to look for image files as specified by the
// image attribute or using the IMG element in HTML-like labels.
// The string should be a list of (absolute or relative)
// pathnames, each separated by a semicolon (for Windows)
// or a colon (all other OS).
func ImagesPath(paths string) GraphOption {
	return func(gr *dot.Graph) {
		gr.Attr("imagepath", paths)
	}
}

// Vertical enables/diables the top to bottom layout direction.
func Vertical(set bool) GraphOption {
	return func(gr *dot.Graph) {
		if set {
			gr.Attr("rankdir", "TB")
		} else {
			gr.Attr("rankdir", "LR")
		}
	}
}

// newGraph returns a new GraphViz DOT language graph
func newGraph(opts ...GraphOption) *dot.Graph {
	res := dot.NewGraph(dot.Undirected)

	res.Attr("rankdir", "LR")
	res.Attr("pad", "1")
	res.Attr("ranksep", "2.3")
	res.Attr("nodesep", "0.8")
	res.Attr("fontname", "Fira Code")
	res.Attr("fontsize", "14")
	res.Attr("splines", "curved")
	res.Attr("concentrate", "true")
	res.Attr("orientation", "portrait")

	for _, opt := range opts {
		opt(res)
	}

	return res
}

// createEdge creates a new connection line between two nodes
func createEdge(gr *dot.Graph, fid, tid string, color string) error {
	a, ok := gr.FindNodeById(fid)
	if !ok {
		return fmt.Errorf("node with id=%s not found", fid)
	}

	b, ok := gr.FindNodeById(tid)
	if !ok {
		return fmt.Errorf("node with id=%s not found", tid)
	}

	res := gr.Edge(a, b)

	res.Attr("fontname", "Fira Code")
	res.Attr("fontsize", "10")
	res.Attr("penwidth", "2.5")
	//res.Attr("xlabels", strconv.Itoa(lvl))

	if strings.TrimSpace(color) != "" {
		res.Attr("color", color)
	} else {
		res.Attr("color", "#ced4da")
	}

	return nil
}

// createNode create and adds a new node to the graph.
// You can customize some attributes using
// the variadic node attributes.
func createNode(g *dot.Graph, id string, opts ...nodeAttribute) string {
	n := g.Node(id)
	n.Label("")
	n.Attr("fontname", "Fira Code")
	n.Attr("fontsize", "12")
	n.Attr("width", "2")
	n.Attr("margin", "0.2,0.2")
	n.Attr("shape", "plain")

	for _, opt := range opts {
		opt(&n)
	}

	return id
}

// nodeAttribute defines a function that
// apply a property to a node
type nodeAttribute func(*dot.Node)

// nodeFillColor sets the node fill color
func nodeFillColor(hex string) nodeAttribute {
	return func(el *dot.Node) {
		if strings.TrimSpace(hex) == "" {
			return
		}

		el.Attr("fillcolor", hex)

		const attr = "filled"

		val := el.AttributesMap.Value("style")
		if val == nil {
			el.Attr("style", attr)
			return
		}

		style, ok := val.(string)
		if !ok {
			el.Attr("style", attr)
			return
		}

		if _, found := text.Find(strings.Split(style, ","), attr); !found {
			el.Attr("style", strings.Join([]string{style, attr}, ","))
		}
	}
}

// nodeFontSize specify the font size, in points, used for text
func nodeFontSize(s int) nodeAttribute {
	return func(el *dot.Node) {
		el.Attr("fontsize", strconv.Itoa(s))
	}
}

// nodeLabel is the node caption
// if 'htm' is true the caption is treated as HTML code
func nodeLabel(label string, htm bool) nodeAttribute {
	return func(el *dot.Node) {
		if htm {
			el.Attr("label", dot.HTML(label))
		} else {
			el.Attr("label", label)
		}
	}
}

// nodeShape sets the shape of a node
func nodeShape(shape string) nodeAttribute {
	return func(el *dot.Node) {
		el.Attr("shape", shape)
	}
}
